src/query/api/v1/handler/prom/read.go

```diff
@@ -109,13 +134,69 @@ func newReadHandler(
 	scope := hOpts.InstrumentOpts().MetricsScope().Tagged(
 		map[string]string{"handler": "prometheus-read"},
 	)
-	return &readHandler{
+	var qs *queryShadowing = nil
+	if hOpts.ShadowQueryURL() != "" {
+		var err error
+		qs, err = newQueryShadowing(hOpts.ShadowQueryURL(), hOpts.QueryShadowingWorkers())
+		if err != nil {
+			return nil, err
+		}
+	}
+	handler := &readHandler{
 		hOpts:               hOpts,
 		opts:                options,
 		scope:               scope,
 		logger:              hOpts.InstrumentOpts().Logger(),
 		returnedDataMetrics: native.NewPromReadReturnedDataMetrics(scope),
-	}, nil
+		qs: 			     qs,
+	}
+	if handler.qs != nil {
+		handler.logger.Info("Query shadowing is enabled",
+		    zap.String("shadowQueryURL", handler.qs.shadowQueryURL),
+			zap.Int("QueryShadowingWorkers", hOpts.QueryShadowingWorkers()),
+		)
+	}
+	return handler, nil
+}
+
+func (h* readHandler) sendShadowQuery(r *http.Request) {
```

Comment:
we shall have the metrics to record success/error as well as latency. we can leave the correctness validation to rule evaluation only using structure logging.

Reply:
I'm planned to add success/error counters, but measuring latency doesn't help much because of the extra hop to Pantheon. The latency here won't be an apple-to-apple comparison with M3. Latency can be measured on Pantheon side.